Tests/Test.cs

```diff
@@ -0,0 +1,321 @@
+using System.Collections.Immutable;
+using FluentAssertions;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Time.Testing;
+
+namespace Tests;
+
+using static ShoppingCartCommand;
+using static ShoppingCartEvent;
+using static ShoppingCart;
+
+// Types Aliases
+using ProductItems = ImmutableDictionary<string, int>;
+using ProductItem = (Guid ProductId, int Quantity);
+using PricedProductItem = (Guid ProductId, int Quantity, decimal Price);
+
+// Commands
+public abstract record ShoppingCartCommand
+{
+    public record AddProduct(
+        ProductItem ProductItem,
+        Guid? ClientId = null
+    ): ShoppingCartCommand;
+
+    public record RemoveProduct(
+        PricedProductItem ProductItem
+    ): ShoppingCartCommand;
+
+    public record Confirm(
+        Guid ClientId
+    ): ShoppingCartCommand;
+
+    public record Cancel: ShoppingCartCommand;
+
+    private ShoppingCartCommand() { }
+}
+
+public abstract record ShoppingCartEvent
+{
+    public record Opened(
+        Guid? ClientId,
+        DateTimeOffset OpenedAt
+    ): ShoppingCartEvent;
+
+    public record ProductAdded(
+        PricedProductItem ProductItem,
+        DateTimeOffset AddedAt
+    ): ShoppingCartEvent;
+
+    public record ProductRemoved(
+        PricedProductItem ProductItem,
+        DateTimeOffset RemovedAt
+    ): ShoppingCartEvent;
+
+    public record Confirmed(
+        Guid ClientId,
+        DateTimeOffset ConfirmedAt
+    ): ShoppingCartEvent;
+
+    public record Cancelled(
+        DateTimeOffset CanceledAt
+    ): ShoppingCartEvent;
+
+    private ShoppingCartEvent() { }
+}
+
+public abstract record ShoppingCart
+{
+    public record Empty: ShoppingCart;
+
+    public record Pending(ProductItems ProductItems): ShoppingCart;
+
+    public record Closed: ShoppingCart;
+
+    private ShoppingCart() { }
+}
+
+// Primary constructor & TimeProvider
+// Then Collection expressions and advanced pattern matching
+public class ShoppingCartDecider(
+    IProductPriceCalculator priceCalculator,
+    TimeProvider timeProvider
+)
+{
+    public ShoppingCartEvent[] Decide(ShoppingCartCommand command, ShoppingCart state) =>
```

Comment:
Minor thing: I'd switch the order of parameters here to => ShoppingCart state, ShoppingCartCommand command

Reply:
I wouldn't - the state is the thing that is obtained last. It varies if there is a conflict and a retry is required. Per normal partial application, invocation of a command would have some known inputs, potentially decorated with authentication/claims data, then some more info from inner layers - the state is the final piece of the puzzle.

(In an F# decider impl, the state goes first [and the event goes last] in an `evolve` function; that's the main exception)